### ISSUE TITLE
Typeguard Error since 4.9 with unrelated generic

### ISSUE THREAD

**Issue by hasezoey** [microsoft/TypeScript#51638]
# Bug Report

Since typescript 4.9 a typeguard with multiple generics and a conditional assertion (`val is`) throws a error with a unrelated generic

### 🔎 Search Terms

4.9 typeguard conditional return 4.8

### 🕗 Version & Regression Information

- This is a regression
- This changed between versions 4.8 and 4.9
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground Typescript 4.8.4](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5xBAdgcwhVAUzgDMoIQ4AiTXfIqgbgChmYBPMYgJUJIB4AChDABXADbIYhACYAVToQA0cbsgDuASRlxCAD2lYZ6WngKEAdLxIKuAPjgBeDNjNELwsZOkyAIhABjIREJKVlbZVUNbTsWDi5VPginFzpzKyTFOMU4fwDREEIsGAj+ORUARVFCKHYACUJxLih0ZwBvAF8HZ1N6SzyCoph+ZCx2Surahqaa1BU5BwAyODkWZhJRLACYYGw4YFQBwuKylQBlXQMi40SbRTsACmY4OBlAgC5b07gzu2YASk+bwC+3QckuhhuRyGpWAWBINRWAEYVHCEVAVgAmBwAflygUGxVKchRWIcn3B+kh6AgACMAFaEbZwPHQomKMrkuBYQgAN0RbWecCghBgoigWFegX2WFQMFGAUIEBIqTclgAshAZI0WB0gA)

[Playground Typescript next (4.9 not available)](https://www.typescriptlang.org/play?ts=5.0.0-dev.20221124#code/JYWwDg9gTgLgBAKjgQwM5xBAdgcwhVAUzgDMoIQ4AiTXfIqgbgChmYBPMYgJUJIB4AChDABXADbIYhACYAVToQA0cbsgDuASRlxCAD2lYZ6WngKEAdLxIKuAPjgBeDNjNELwsZOkyAIhABjIREJKVlbZVUNbTsWDi5VPginFzpzKyTFOMU4fwDREEIsGAj+ORUARVFCKHYACUJxLih0ZwBvAF8HZ1N6SzyCoph+ZCx2Surahqaa1BU5BwAyODkWZhJRLACYYGw4YFQBwuKylQBlXQMi40SbRTsACmY4OBlAgC5b07gzu2YASk+bwC+3QckuhhuRyGpWAWBINRWAEYVHCEVAVgAmBwAflygUGxVKchRWIcn3B+kh6AgACMAFaEbZwPHQomKMrkuBYQgAN0RbWecCghBgoigWFegX2WFQMFGAUIEBIqTclgAshAZI0WB0gA)

### 💻 Code

```ts
import * as mongoose from "mongoose";

type Ref<PopulatedType, RawId extends mongoose.RefType> = mongoose.PopulatedDoc<PopulatedType, RawId>;
type RefType = mongoose.RefType;
type DocumentType<T, QueryHelpers = {}> = mongoose.Document<any, QueryHelpers, T> & T;

function isDocument<T, S extends RefType>(
  doc: Ref<T, S>
): doc is T extends DocumentType<infer T1, infer T2> ? DocumentType<T1, T2> : T extends object ? DocumentType<T> : never {
  return doc instanceof mongoose.Model;
}
```

### 🙁 Actual behavior

in 4.9 (and next), Error:

```
A type predicate's type must be assignable to its parameter's type.
  Type 'T extends DocumentType<infer T1, infer T2> ? DocumentType<T1, T2> : T extends object ? DocumentType<T, {}> : never' is not assignable to type 'Ref<T, S>'.
    Type 'Document<any, unknown, unknown> | (T extends object ? DocumentType<T, {}> : never)' is not assignable to type 'Ref<T, S>'.
      Type 'Document<any, unknown, unknown>' is not assignable to type 'Ref<T, S>'.
        Type 'Document<any, unknown, unknown>' is not assignable to type 'S'.
          'S' could be instantiated with an arbitrary type which could be unrelated to 'Document<any, unknown, unknown>'.
```

### 🙂 Expected behavior

No Error (like in 4.8)

---

from what i understand, generic `S` has nothing to do with the return type

**Comment by MartinJohns** [comment]
> from what i understand, generic `S` has nothing to do with the return type

But it does. The return type is a type guard, and the argument must be assignable to the type guard, and the type of the argument includes `S`.

**Comment by RyanCavanaugh** [comment]
This error looks correct to me. We'd need an isolated repro clearly demonstrating a defect -- we don't have the time to dig through all the Mongoose types to understand what's going on here without a good case that something has gone wrong.

**Comment by doanestudio** [comment]
[](https://github.com/typegoose/typegoose/issues/772#issuecomment-1334233320)

**Comment by doanestudio** [comment]
> 🎉 This issue has been resolved in version 10.0.0-beta.1 🎉
> 
> The release is available on:
> 
> * [npm package (@beta dist-tag)](https://www.npmjs.com/package/@typegoose/typegoose/v/10.0.0-beta.1)
> * [GitHub release](https://github.com/typegoose/typegoose/releases/tag/v10.0.0-beta.1)
> 
> Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot 📦🚀
